File: 3.longest-substring-without-repeating-characters.js
// @leet start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
	if (s.length<=1) {
		return s.length
	}
	var hash = new Map()
	var answer = 0;
	var left = 0;
	var right = 0;
	while (right<s.length) {
		if (!hash.has(s[right])) {
			hash.set(s[right],true)
		}else{
			while (s[left]!==s[right]) {
				hash.delete(s[left])
				left++;
			}
			left++;
		}
		answer = Math.max(answer,right-left+1)
		right++;
	}
	return answer
};
// @leet end

File: 3.longest-substring-without-repeating-characters.cpp
// @leet start
#include <string>
#include <unordered_map>
#include <cstdlib>
using namespace std;
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
		if (s.size()==0) {
			return 0;
		}
    unordered_map<char, int> map;
    int maxLength = 0;
    int start = 0;
		for (int i = 0; i < s.size(); i++) {
			if (map.find(s[i]) != map.end()) {
				start = max(start,map[s[i]]+1);
			}
			map[s[i]]=i;
			maxLength = max(maxLength, i-start+1);
		}
    return maxLength;
  }
};
// @leet end
